pkg/controller/networkpolicy/clustergroup.go

```diff
@@ -176,27 +173,62 @@ func (n *NetworkPolicyController) syncInternalGroup(key string) error {
 		klog.V(2).Infof("Internal group %s not found.", key)
 		return nil
 	}
-
 	grp := grpObj.(*antreatypes.Group)
-	// Find all Pods matching its selectors and update store.
-	groupSelector := grp.Selector
-	pods, _ := n.processSelector(groupSelector)
-	memberSet := controlplane.GroupMemberSet{}
-	for _, pod := range pods {
-		if pod.Status.PodIP == "" {
-			// No need to insert Pod IPAddress when it is unset.
-			continue
+	if grp.IPBlock == nil {
+		// Find all Pods matching its selectors and update store.
+		groupSelector := grp.Selector
+		pods, _ := n.processSelector(groupSelector)
+		memberSet := controlplane.GroupMemberSet{}
+		for _, pod := range pods {
+			if pod.Status.PodIP == "" {
+				// No need to insert Pod IPAddress when it is unset.
+				continue
+			}
+			memberSet.Insert(podToGroupMember(pod, true))
 		}
-		memberSet.Insert(podToGroupMember(pod, true))
+		// Update the internal Group object in the store with the Pods as GroupMembers.
+		updatedGrp := &antreatypes.Group{
+			UID:          grp.UID,
+			Name:         grp.Name,
+			Selector:     grp.Selector,
+			GroupMembers: memberSet,
+		}
+		klog.V(2).Infof("Updating existing internal Group %s with %d GroupMembers", key, len(memberSet))
+		n.internalGroupStore.Update(updatedGrp)
+	}
+	// Retrieve the ClusterGroup corresponding to this key.
+	cg, err := n.cgLister.Get(grp.Name)
+	if err != nil {
+		klog.Infof("Didn't find the ClusterGroup %s, skip processing of internal group", grp.Name)
+		return nil
+	}
+	// If a ClusterGroup is added/updated, it might have a reference in ClusterNetworkPolicy.
+	cnps, err := n.cnpInformer.Informer().GetIndexer().ByIndex(ClusterGroupIndex, cg.Name)
+	if err != nil {
+		klog.Errorf("Error retrieve ClusterNetworkPolicies corresponding to ClusterGroup %s", cg.Name)
+		return err
 	}
-	// Update the internal Group object in the store with the Pods as GroupMembers.
-	updatedGrp := &antreatypes.Group{
-		UID:          grp.UID,
-		Name:         grp.Name,
-		Selector:     grp.Selector,
-		GroupMembers: memberSet,
+	for _, obj := range cnps {
+		cnp := obj.(*secv1alpha1.ClusterNetworkPolicy)
+		// Re-process ClusterNetworkPolicies which may be affected due to updates to CG.
+		curInternalNP := n.processClusterNetworkPolicy(cnp)
```

Comment:
Do we really need a CNP re-sync here? `syncInternalGroup` is called whenever updates to a Group happens. If CNPs refers to this group, we only need to make sure that addressGroup.members are updated. And since addressGroup and its corresponded internalGroup share the same uid, that should not be complicated. The internalNP has nothing to do with these updates here. Updates to CNP spec (group references) will be taken care of by add/update/deleteCNP functions. I don't know if there's something that I miss here.

Reply:
think of the scenario when the CG is updated to remove the (ns)podSelector and replaced with an IPBlock and that CG was originally referenced in a NP.
The NP AddressGroup must now be deleted. The NP IPBlocks must now be updated.

Reply:
Got it. Somehow I remembered ipBlock was a part of addressGroup. This makes sense to me.